spec/module_validator.rb

```diff
@@ -0,0 +1,151 @@
+require 'active_model'
+
+class ArrayInclusionValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    unless value.is_a?(Array)
+      record.errors.add(attribute, "#{attribute} must be an array")
+      return
+    end
+
+    if attribute == :references
+      reference_keys = []
+      value.each { |k, _v| reference_keys << k }
+      value = reference_keys
+    end
+
+    invalid_options = value - options[:in]
+    message = "contains invalid values #{invalid_options.inspect} - only #{options[:in].inspect} is allowed"
+
+    if invalid_options.any?
+      record.errors.add(attribute, :array_inclusion, message: message, value: value)
+    end
+  end
+end
+
+class ModuleValidator < SimpleDelegator
+  include ActiveModel::Validations
+
+  validate :validate_filename_is_snake_case, :validate_reference_ctx_id, :validate_author_bad_chars
+
+  attr_reader :mod
+
+  def initialize(mod)
+    super
+    @mod = mod
+  end
+
+  #
+  # Acceptable Stability ratings
+  #
+  VALID_STABILITY_VALUES = [
```

Comment:
Wondering if we need a reference here https://github.com/rapid7/metasploit-framework/blob/1bd3a764a67fe9beb4123e300e3bb83d0f4ffa2c/lib/msf/core/constants.rb#L59 mentioning that any values added need to be made here as well, be nicer if we could just have a single place where these are defined but I don't know if we can do that without changing namespacing

Reply:
I believe @adfoster-r7 and I discussed this and landed on just adding them here instead, as I originally added them to `constants.rb` and called off to that in my tests. I don't remember exactly why we did that off hand if I'm honest.

Reply:
I'm not strongly opinionated here; I'd prefer the constants to have been namespaced way-back-when, but that's not worth the effort fixing now.

If someone adds a new constant, it will break this new validation logic anyways - so I don't think we need the pointer back to this file (famous last words)